commands/verify/verify.go

```diff
@@ -3,93 +3,90 @@ package verify
 import (
 	"context"
 	"fmt"
-	"os"
 	"path/filepath"
-	
-	"github.com/instrumenta/conftest/policy"
+
 	"github.com/instrumenta/conftest/commands/test"
-	"github.com/containerd/containerd/log"
+	"github.com/instrumenta/conftest/policy"
 	"github.com/open-policy-agent/opa/tester"
 	"github.com/spf13/cobra"
 	"github.com/spf13/viper"
 )
 
-func NewVerifyCommand(getOutputManager func() test.OutputManager) *cobra.Command {
-	ctx := context.Background()
-	cmd := &cobra.Command{
+func NewVerifyCommand(ctx context.Context) *cobra.Command {
+	cmd := cobra.Command{
 		Use:   "verify",
 		Short: "Verify Rego unit tests",
-		PreRun: func(cmd *cobra.Command, args []string) {
+		PreRunE: func(cmd *cobra.Command, args []string) error {
 			err := viper.BindPFlag("output", cmd.Flags().Lookup("output"))
 			if err != nil {
-				log.G(ctx).Fatal("Failed to bind argument:", err)
+				return fmt.Errorf("bind output: %w", err)
 			}
-		},
-		Run: func(cmd *cobra.Command, args []string) {
-			out := getOutputManager()
 
+			return nil
+		},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			outputManager := test.GetOutputManager()
 			policyPath := viper.GetString("policy")
-			compiler, err := policy.BuildCompiler(policyPath, true)
-			if err != nil {
-				log.G(ctx).Fatalf("Problem building rego compiler: %s", err)
-			}
-
-			runner := tester.NewRunner().
-				SetCompiler(compiler)
 
-			ch, err := runner.Run(ctx, compiler.Modules)
+			results, err := RunVerification(ctx, policyPath)
 			if err != nil {
-				log.G(ctx).Fatalf("Problem running rego tests: %s", err)
+				return fmt.Errorf("running verification: %w", err)
 			}
 
-			results := getResults(ctx, ch)
-
-			for result := range results {
-				msg := fmt.Errorf("%s", result.Msg)
-				fileName := filepath.Join(policyPath, result.FileName)
-				if result.Fail {
-					err = out.Put(fileName, test.CheckResult{Failures: []error{msg}})
-				} else {
-					err = out.Put(fileName, test.CheckResult{Successes: []error{msg}})
-				}
-
-				if err != nil {
-					log.G(ctx).Fatalf("Problem writing to output: %s", err)
+			for _, result := range results {
+				if err := outputManager.Put(result.FileName, result); err != nil {
+					return fmt.Errorf("put result: %w", err)
 				}
 			}
 
-			err = out.Flush()
-			if err != nil {
-				log.G(ctx).Fatal(err)
+			if err := outputManager.Flush(); err != nil {
+				return fmt.Errorf("flushing output: %w", err)
 			}
 
-			os.Exit(0)
+			return nil
 		},
 	}
 
 	cmd.Flags().StringP("output", "o", "", fmt.Sprintf("output format for conftest results - valid options are: %s", test.ValidOutputs()))
 
-	return cmd
+	return &cmd
 }
 
-type report struct {
-	FileName string
-	Msg      string
-	Fail     bool
-}
+// RunVerification runs the verify command
+func RunVerification(ctx context.Context, path string) ([]test.CheckResult, error) {
```

Comment:
I really like the separation this provides. 
Can we wrap some tests around it to guard against regressions and to stress all of its behaviors?

Reply:
I'm on the fence about adding _unit_ tests specifically right now. We do already have coverage with `acceptance.bats`.

The reason why I waver is that I'm not a huge fan of the `testdata` folders in the commands, and this function would need something similar, a `.rego` that exists on disk.

I really want to be able to define a `Rego` file as a `[]byte` and pass it in, rather than depending on `path`. But that means allowing `policy.BuildCompiler` to also take a `Reader` rather than depending on path, and all the way down. We have too much reliance on `.Open` and `.ReadAll` right now.

I envision us eventually/soon moving to a "Conftest" sort of `struct` that has a compiler, config, etc where we have more control over the dependencies, giving us more flexibility with tests.

Reply:
@jpreese 
I agree with your sentiments around adding more `testdata` files, and I too would love to see us more more towards using Readers instead of passing around file path strings. 

My real concern is that this is an exported function and as such could be called directly by a user of the package, so our current outside in approach is a layer too high to provide any guarantees around this function's behaviors. What are your thoughts on just making this function un-exported, since I don't think we want anyone calling this directly anyway, and therefore no unit test for this function is needed, as the only public API of note to test is at the command level, which is covered by our acceptance tests. 

Does that make sense?

Reply:
I'm OK with that. I don't really know what our public-facing package structure/methods are going to look like yet, so we can keep it private until we have a clearer vision.

As I worked through this, and starting taking note of what depends on what, I imagine `verify` and `test` are going to be combined into a single package.